java/java-runtime/src/com/intellij/rt/debugger/MetadataDebugHelper.java

```diff
@@ -0,0 +1,146 @@
+   *
+   * The metadata is returned as a string in JSON format, where the `data1` field is
+   * base64 encoded to keep the format valid.
+   *
+   * Returning metadata as a single string allows us to minimize the number of JDI calls from
+   * the debugger side. Any JDWP communication can be very expensive when debugging remotely,
+   * especially when debugging Android applications on a phone.
+   */
+  @SuppressWarnings("unchecked")
+  public static String getDebugMetadataAsJson(Class<?> cls) {
+    try {
+      Class<?> metadataClass = Class.forName(METADATA_CLASS_NAME, false, cls.getClassLoader());
+      Method kindMethod = metadataClass.getDeclaredMethod(KIND_METHOD_NAME);
+      Method metadataVersionMethod = metadataClass.getDeclaredMethod(METADATA_VERSION_METHOD_NAME);
+      Method data1Method = metadataClass.getDeclaredMethod(DATA_1_METHOD_NAME);
+      Method data2Method = metadataClass.getDeclaredMethod(DATA_2_METHOD_NAME);
+      Method extraStringMethod = metadataClass.getDeclaredMethod(EXTRA_STRING_METHOD_NAME);
+      Method packageNameMethod = metadataClass.getDeclaredMethod(PACKAGE_NAME_METHOD_NAME);
+      Method extraIntMethod = metadataClass.getDeclaredMethod(EXTRA_INT_METHOD_NAME);
+
+      Object metadata = cls.getAnnotation((Class<Annotation>)metadataClass);
+
+      StringBuilder sb = new StringBuilder();
+      sb.append("{");
+
+      int kind = (int)kindMethod.invoke(metadata);
+      appendAsJsonValue(sb, KIND, kind);
+
+      int[] metadataVersion = (int[]) metadataVersionMethod.invoke(metadata);
+      appendAsJsonValue(sb, METADATA_VERSION, Arrays.toString(metadataVersion));
+
+      String[] data1 = (String[])data1Method.invoke(metadata);
+      appendAsJsonValue(sb, DATA_1, Arrays.toString(toJsonBase64Encoded(data1)));
+
+      String[] data2 = (String[])data2Method.invoke(metadata);
+      appendAsJsonValue(sb, DATA_2, Arrays.toString(toJson(data2)));
+
+      String extraString = (String) extraStringMethod.invoke(metadata);
+      appendAsJsonValue(sb, EXTRA_STRING, toJson(extraString));
+
+      String packageName = (String) packageNameMethod.invoke(metadata);
+      appendAsJsonValue(sb, PACKAGE_NAME, toJson(packageName));
+
+      int extraInt = (int)extraIntMethod.invoke(metadata);
+      appendAsJsonValueNoComma(sb, EXTRA_INT, extraInt);
+
+      sb.append("}");
+
+      return sb.toString();
+    } catch (Exception ex) {
+      return null;
+    }
+  }
+
+  /*
+   * This function is used similarly to `getDebugMetadataAsJson`, when there is a need to
+   * fetch metadata for multiple classes.
+   *
+   * The return value is a concatenation of metadata JSON representations of given classes
+   * separated by MetadataDebugHelper.METADATA_SEPARATOR.
+   */
+  public static String getDebugMetadataListAsJson(Class<?>... classes) {
+    StringBuilder sb = new StringBuilder();
+    for (Class<?> cls : classes) {
+      String metadataAsJson = getDebugMetadataAsJson(cls);
```

Comment:
It could be minor, but what do you think about creating an additional private method, where we can pass StringBuilder?
We can reuse it then, instead of creating a new string every time